<a href="https://colab.research.google.com/github/julianavtal/pandas-essencial/blob/main/manipulacao_de_dados_codigo_base.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Pandas Essencial**
**Prof. Dr. Samuel Martins (@hisamuka @xavecoding)** <br/>
xavecoding: https://youtube.com/c/xavecoding <br/>

Neste tutorial, vamos aprender o essencial da biblioteca _Pandas_ para manipulação de dados.<br/><br/>

<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.

### Pacotes usados neste Notebook

In [102]:
# pacotes usados neste notebook
import pandas as pd

<h1>Manipulação de Dados com Pandas</h1><hr/>

<h2>1. Manipulação Básica de Datasets</h2>
<hr/>

**Dataset**: Gas Prices in Brazil: https://www.kaggle.com/matheusfreitag/gas-prices-in-brazil <br/>

Este dataset contém os **registros dos preços médios semanais dos combustíveis do Brasil entre os anos de 2004 e 2019**. <br/>
Cada *amostra (registro/linha)* consiste em um registro de preço aferido para um dado tipo de combustível em uma dada localidade do Brasil. <br/>
Alguns dos principais *atributos* (colunas) do dataset são: 'ESTADO', 'PRODUTO', 'NÚMERO DE POSTOS PESQUISADOS', 'PREÇO MÉDIO REVENDA'.


\* O arquivo disponibilizado no Kaggle está no formato *tsv*. Embora o _pandas_ consiga abrí-lo normalmente, convertemos tal arquivo para o formato *CSV*, que é um dos formatos mais utilizados, e mudamos seu separado para ';' apenas para mostrar algumas opções da função de carregamento.

### 1.1. Importando o Dataset
Para carregar um dataset no formato csv, basta utilizar a função `read_csv` do pandas. Por padrão, ela considera _','_ como separador.

In [103]:
import google.colab.drive as drive
drive.mount('/content/drive')
df = pd.read_csv('/content/drive/MyDrive/Data Science 2025 IFSP/datasets/GasPricesinBrazil_2004-2019.csv')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [104]:
df

,Unnamed: 0;DATA INICIAL;DATA FINAL;REGIÃO;ESTADO;PRODUTO;NÚMERO DE POSTOS PESQUISADOS;UNIDADE DE MEDIDA;PREÇO MÉDIO REVENDA;DESVIO PADRÃO REVENDA;PREÇO MÍNIMO REVENDA;PREÇO MÁXIMO REVENDA;MARGEM MÉDIA REVENDA;COEF DE VARIAÇÃO REVENDA;PREÇO MÉDIO DISTRIBUIÇÃO;DESVIO PADRÃO DISTRIBUIÇÃO;PREÇO MÍNIMO DISTRIBUIÇÃO;PREÇO MÁXIMO DISTRIBUIÇÃO;COEF DE VARIAÇÃO DISTRIBUIÇÃO;MÊS;ANO
0,0;2004-05-09;2004-05-15;CENTRO OESTE;DISTRITO ...
1,1;2004-05-09;2004-05-15;CENTRO OESTE;GOIAS;ETA...
2,2;2004-05-09;2004-05-15;CENTRO OESTE;MATO GROS...
3,3;2004-05-09;2004-05-15;CENTRO OESTE;MATO GROS...
4,4;2004-05-09;2004-05-15;NORDESTE;ALAGOAS;ETANO...
...,...
106818,106818;2019-06-23;2019-06-29;NORDESTE;RIO GRAN...
106819,106819;2019-06-23;2019-06-29;SUL;RIO GRANDE DO...
106820,106820;2019-06-23;2019-06-29;SUL;SANTA CATARIN...
106821,106821;2019-06-23;2019-06-29;SUDESTE;SAO PAULO...


O dataset não foi carregado corretamente pois o separador utilizado seu arquivo CSV era ';' e não a ','. <br/>
Vamos então carregá-lo corretamente:

In [105]:
df = pd.read_csv('/content/drive/MyDrive/Data Science 2025 IFSP/datasets/GasPricesinBrazil_2004-2019.csv', sep=';')

In [106]:
df

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,0.065,...,3.690,0.818,0.018,2.756,0,2.756,2.756,0,6,2019
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,0.129,...,3.789,0.95,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,0.190,...,3.499,0.914,0.065,1.998,0,1.9981,1.9981,0,6,2019
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,0.229,...,3.490,0.646,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019


### 1.2. Exibindo as primeiras linhas do Dataset
A função `.head()` exibe as 5 primeiras linhas do dataset/tabela/Data Frame.

In [107]:
df.head()

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


In [108]:
df.head(1)

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.35,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004


### 1.3 Informações do Dataset e Elementos Chave

### 1.3.1 Informações gerais sobre o Dataset

In [109]:
#função info() explica os tipos de dado pra cada uma das colunas
#tipo 'object' no geral representa strings

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 106823 entries, 0 to 106822
Data columns (total 21 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Unnamed: 0                     106823 non-null  int64  
 1   DATA INICIAL                   106823 non-null  object 
 2   DATA FINAL                     106823 non-null  object 
 3   REGIÃO                         106823 non-null  object 
 4   ESTADO                         106823 non-null  object 
 5   PRODUTO                        106823 non-null  object 
 6   NÚMERO DE POSTOS PESQUISADOS   106823 non-null  int64  
 7   UNIDADE DE MEDIDA              106823 non-null  object 
 8   PREÇO MÉDIO REVENDA            106823 non-null  float64
 9   DESVIO PADRÃO REVENDA          106823 non-null  float64
 10  PREÇO MÍNIMO REVENDA           106823 non-null  float64
 11  PREÇO MÁXIMO REVENDA           106823 non-null  float64
 12  MARGEM MÉDIA REVENDA          

A primeira coluna da tabela, chamada _'Unnamed: 0'_, parece não significar nada. Na verdade, ela parece ser os **índices** da tabela, que foram salvos como uma _coluna_.<br/>
Veremos jajá como **remover tal coluna**.

Outro ponto é que, aparentemente, nenhum atributo/coluna possui valores nulos (_null_), uma vez que o número de registros do dataframe e os números de valores _non-null_ é de **106823**. <br/>
Mas, veremos que não é bem assim para esse caso.

### 1.3.2 Data Frame
Todo dataset carregado (dados estruturados) é um `Data Frame`: 'Tabela' bi-dimensional, de tamanho mutável, com dados potencialmente heterogêneos. <br/>

In [110]:
type(df)

pandas.core.frame.DataFrame

Podemos acessar as **dimensões do Data Frame** (número de linhas x número de colunas) utilizando o atributo `.shape` do Data Frame.

In [111]:
#dúvida conceitual: o que é um atributo em py? o que é um método? qual a diferença pra uma função
df.shape

#o resultado é uma tupla, podemos acessar por index

(106823, 21)

In [112]:
#o f' funciona para usarmos {} e não precisar concatenar strings
# usar o \n para pular pra próxima linha

print(f'número de linhas: {df.shape[0]} \nnúmero de colunas: {df.shape[1]}')

número de linhas: 106823 
número de colunas: 21


#### **Criando um DataFrame**

Podemos criar um DataFrame a partir de um _dicionário_, onde cada **chave** possui uma **lista de elementos de igual tamanho**.<br/>
As **chaves** representam as **colunas** e **cada um dos valores de sua lista** representa o **valor da linha** correspondente para aquela coluna.

In [113]:
#criando um df a partir de um dicionário
personagens_df = pd.DataFrame({
    'nome': ['Luke Skywalker', 'Yoda', 'Palpatine'],
    'idade': [16, 1000, 70],
    'peso': [70, 15, 60],
    'eh jedi': [True, True, False]  # o nome das colunas podem ter espaços
})

In [114]:
personagens_df

,nome,idade,peso,eh jedi
0,Luke Skywalker,16,70,True
1,Yoda,1000,15,True
2,Palpatine,70,60,False


In [115]:
personagens_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   nome     3 non-null      object
 1   idade    3 non-null      int64 
 2   peso     3 non-null      int64 
 3   eh jedi  3 non-null      bool  
dtypes: bool(1), int64(2), object(1)
memory usage: 207.0+ bytes


#### **VEJA MAIS**
Criando um Data Frame a partir de um dicionário: https://www.geeksforgeeks.org/how-to-create-dataframe-from-dictionary-in-python-pandas/

#### **Renomeando as colunas de um DataFrame**
**===>** O atributo `DataFrame.columns` retorna uma "lista" com os **nomes de todas as colunas** do data frame.

In [116]:
personagens_df.columns

Index(['nome', 'idade', 'peso', 'eh jedi'], dtype='object')

In [117]:
type(personagens_df.columns)

pandas.core.indexes.base.Index

In [118]:
#transformando as colunas em uma lista
list(personagens_df.columns)

['nome', 'idade', 'peso', 'eh jedi']

<br/>

**===>** Para **renomear colunas** do data frame, utilize o método `DataFrame.rename`, que retorna uma _cópia_ do data frame com as as colunas renomeadas:

In [119]:
personagens_df

,nome,idade,peso,eh jedi
0,Luke Skywalker,16,70,True
1,Yoda,1000,15,True
2,Palpatine,70,60,False


In [120]:
#para renomear, passamos um dicionário

personagens_df.rename(columns = {
                    'nome' : "Nome"
                    ,"idade" : "Idade"
                    ,'peso' : 'Peso'
                    ,'eh jedi' : 'É Jedi?'

                    })

,Nome,Idade,Peso,É Jedi?
0,Luke Skywalker,16,70,True
1,Yoda,1000,15,True
2,Palpatine,70,60,False


In [121]:
#para as mudanças serem salvas no df, necessário usar o inplace
personagens_df

,nome,idade,peso,eh jedi
0,Luke Skywalker,16,70,True
1,Yoda,1000,15,True
2,Palpatine,70,60,False


Para renomear o _próprio_ data frame em questão, utilize o parâmetro `inplace=True`:

In [122]:
#usando inplace para salvar as mudanças no df

personagens_df.rename(columns = {
                    'nome' : "Nome"
                    ,"idade" : "Idade"
                    ,'peso' : 'Peso'
                    ,'eh jedi' : 'É Jedi?'

                    }, inplace = True)

In [123]:
personagens_df

,Nome,Idade,Peso,É Jedi?
0,Luke Skywalker,16,70,True
1,Yoda,1000,15,True
2,Palpatine,70,60,False


In [124]:
#para arrumar todas de uma vez, por exemplo todas para minusculas
personagens_df.columns = map(str.lower, personagens_df.columns)
personagens_df

,nome,idade,peso,é jedi?
0,Luke Skywalker,16,70,True
1,Yoda,1000,15,True
2,Palpatine,70,60,False


In [125]:
#outra maneira de renomear as colunas é passando os novos nomes via lista
#ou seja precisa ter o mesmo numero de elementos (mesmo número de valores para um número de colunas)

personagens_df.columns = ['Nome', 'Idade','Peso', 'É jedi?']
personagens_df

,Nome,Idade,Peso,É jedi?
0,Luke Skywalker,16,70,True
1,Yoda,1000,15,True
2,Palpatine,70,60,False


In [126]:
#quando é lista, ele já salva automaticamente
personagens_df

,Nome,Idade,Peso,É jedi?
0,Luke Skywalker,16,70,True
1,Yoda,1000,15,True
2,Palpatine,70,60,False


<br/>

**===>** Uma outra forma de **renomear todas as listas** de um data frame é passar uma _lista_ com os novos nomes das colunas para o atributo `DataFrame.columns`:

In [127]:
#exemplo com mismatch = menos valores do que colunas identificadas

#personagens_df.columns = ['Nome', 'Idade','Peso']
#personagens_df

### 1.3.3 Series

Array uni-dimensional com os dados e rótulos de um eixo.

In [128]:
df.head()

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


In [129]:
#selecionando uma coluna toda

df['ESTADO']

,ESTADO
0,DISTRITO FEDERAL
1,GOIAS
2,MATO GROSSO
3,MATO GROSSO DO SUL
4,ALAGOAS
...,...
106818,RIO GRANDE DO NORTE
106819,RIO GRANDE DO SUL
106820,SANTA CATARINA
106821,SAO PAULO


In [130]:
#dúvida: como selecionar determinadas colunas desse jeito?

#ex: df['ESTADO', 'PRODUTO']

In [131]:
#usando head() em uma coluna
df['ESTADO'].head()

,ESTADO
0,DISTRITO FEDERAL
1,GOIAS
2,MATO GROSSO
3,MATO GROSSO DO SUL
4,ALAGOAS


In [132]:
#outra forma de chamar a coluna
#essa forma só funciona pra nomes de coluna sem espaço, caracteres especiais e etc
#aqui é criado um atributo, por isso não pode ter caracteres inválidos
df.ESTADO

,ESTADO
0,DISTRITO FEDERAL
1,GOIAS
2,MATO GROSSO
3,MATO GROSSO DO SUL
4,ALAGOAS
...,...
106818,RIO GRANDE DO NORTE
106819,RIO GRANDE DO SUL
106820,SANTA CATARINA
106821,SAO PAULO


In [133]:
#outra forma de chamar a coluna e usando head()
df.ESTADO.head()

,ESTADO
0,DISTRITO FEDERAL
1,GOIAS
2,MATO GROSSO
3,MATO GROSSO DO SUL
4,ALAGOAS


In [134]:
#da outra maneira
type(df['ESTADO'])

pandas.core.series.Series

In [135]:
#como chamar mais de uma coluna dessa maneira?

In [136]:
#informações sobre uma coluna
type(df.ESTADO)

pandas.core.series.Series

In [137]:
#todos os valores da linha 1

df.iloc[1]

#type sai como 'object' porque existem dados de múltiplos tipos
#dúvida: então o que exatamente é o object?

#importante: com a extração de 1 linha, os índices das colunas são a ordem das colunas (pivot)

,1
Unnamed: 0,1
DATA INICIAL,2004-05-09
DATA FINAL,2004-05-15
REGIÃO,CENTRO OESTE
ESTADO,GOIAS
PRODUTO,ETANOL HIDRATADO
NÚMERO DE POSTOS PESQUISADOS,387
UNIDADE DE MEDIDA,R$/l
PREÇO MÉDIO REVENDA,1.162
DESVIO PADRÃO REVENDA,0.114


#### **Criando uma Series**

Podemos criar um DataFrame a partir de uma lista de elementos.

In [138]:
pd.Series([1.1,3.0,5.2])

,0
0,1.1
1,3.0
2,5.2


Podemos alterar o **nome dos índices** (veremos melhor jajá) e o **nome da Series** (o que ela representa):

In [139]:
pd.Series([1.1,3.0,5.2], index = ['3º colocado', '2º colocado', '1º colocado'])

,0
3º colocado,1.1
2º colocado,3.0
1º colocado,5.2


In [140]:
#dando um nome pra coluna

pd.Series([1.1,3.0,5.2], index = ['3º colocado', '2º colocado', '1º colocado'], name = 'Posição')

,Posição
3º colocado,1.1
2º colocado,3.0
1º colocado,5.2


In [141]:
#classificar como desc
s = pd.Series([1.1,3.0,5.2], index = ['3º colocado', '2º colocado', '1º colocado'])

s.sort_values(ascending=False)

,0
1º colocado,5.2
2º colocado,3.0
3º colocado,1.1


In [142]:
#pra salvar, precisa usar o inplace = True
s

,0
3º colocado,1.1
2º colocado,3.0
1º colocado,5.2


In [143]:
#pra salvar, precisa usar o inplace = True
s.sort_values(ascending=False, inplace=True)
s

,0
1º colocado,5.2
2º colocado,3.0
3º colocado,1.1


#### **VEJA MAIS**
https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html

### 1.3.4 Atribuindo Dados

#### 1.3.4.1 Atribuindo constantes

In [144]:
df.head()

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


In [145]:
df['PRODUTO'] #o retorno não é a coluna em si (uma cópia, uma view), mas sim uma referência à coluna

,PRODUTO
0,ETANOL HIDRATADO
1,ETANOL HIDRATADO
2,ETANOL HIDRATADO
3,ETANOL HIDRATADO
4,ETANOL HIDRATADO
...,...
106818,GNV
106819,GNV
106820,GNV
106821,GNV


In [146]:
#para criar uma cópia, necessário usar o .copy()
df['PRODUTO'].copy()

,PRODUTO
0,ETANOL HIDRATADO
1,ETANOL HIDRATADO
2,ETANOL HIDRATADO
3,ETANOL HIDRATADO
4,ETANOL HIDRATADO
...,...
106818,GNV
106819,GNV
106820,GNV
106821,GNV


In [147]:
#salvando como uma variável

df_produto = df['PRODUTO'].copy()

In [148]:
#se quiser apenas as primeiras linhas
df_produto_head5 = df['PRODUTO'].copy().head()
df_produto_head5

,PRODUTO
0,ETANOL HIDRATADO
1,ETANOL HIDRATADO
2,ETANOL HIDRATADO
3,ETANOL HIDRATADO
4,ETANOL HIDRATADO


In [149]:
#podemos passar um valor constante para toda a coluna
df['PRODUTO'] = 'Combustível'
df.head()

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,Combustível,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,Combustível,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,Combustível,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,Combustível,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,Combustível,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


In [150]:
#recuperando os valores de 'PRODUTO'
df_produto

,PRODUTO
0,ETANOL HIDRATADO
1,ETANOL HIDRATADO
2,ETANOL HIDRATADO
3,ETANOL HIDRATADO
4,ETANOL HIDRATADO
...,...
106818,GNV
106819,GNV
106820,GNV
106821,GNV


#### 1.3.4.2 Atribuindo listas ou series

In [151]:
df.shape

(106823, 21)

In [152]:
nrows = df.shape[0]
nrows

106823

In [153]:
nrows, ncols = df.shape
nrows, ncols

(106823, 21)

In [154]:
produtos_numerados = [f'Produto {i}' for i in range(nrows)]
produtos_numerados

['Produto 0',
 'Produto 1',
 'Produto 2',
 'Produto 3',
 'Produto 4',
 'Produto 5',
 'Produto 6',
 'Produto 7',
 'Produto 8',
 'Produto 9',
 'Produto 10',
 'Produto 11',
 'Produto 12',
 'Produto 13',
 'Produto 14',
 'Produto 15',
 'Produto 16',
 'Produto 17',
 'Produto 18',
 'Produto 19',
 'Produto 20',
 'Produto 21',
 'Produto 22',
 'Produto 23',
 'Produto 24',
 'Produto 25',
 'Produto 26',
 'Produto 27',
 'Produto 28',
 'Produto 29',
 'Produto 30',
 'Produto 31',
 'Produto 32',
 'Produto 33',
 'Produto 34',
 'Produto 35',
 'Produto 36',
 'Produto 37',
 'Produto 38',
 'Produto 39',
 'Produto 40',
 'Produto 41',
 'Produto 42',
 'Produto 43',
 'Produto 44',
 'Produto 45',
 'Produto 46',
 'Produto 47',
 'Produto 48',
 'Produto 49',
 'Produto 50',
 'Produto 51',
 'Produto 52',
 'Produto 53',
 'Produto 54',
 'Produto 55',
 'Produto 56',
 'Produto 57',
 'Produto 58',
 'Produto 59',
 'Produto 60',
 'Produto 61',
 'Produto 62',
 'Produto 63',
 'Produto 64',
 'Produto 65',
 'Produto 66',
 'Pro

In [155]:
#validando se o valor de linhas está batendo
len(produtos_numerados)

106823

In [156]:
#substituindo a coluna PRODUTOS pela lista de 'produtos_numrados'
#necessario que tenha o mesmo numero de valores (no caso, mesmo números de linhas)
df['PRODUTO'] = produtos_numerados
df

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,Produto 0,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,Produto 1,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,Produto 2,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,Produto 3,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,Produto 4,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,Produto 106818,7,R$/m3,3.574,0.065,...,3.690,0.818,0.018,2.756,0,2.756,2.756,0,6,2019
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,Produto 106819,23,R$/m3,3.401,0.129,...,3.789,0.95,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,Produto 106820,24,R$/m3,2.912,0.190,...,3.499,0.914,0.065,1.998,0,1.9981,1.9981,0,6,2019
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,Produto 106821,52,R$/m3,3.020,0.229,...,3.490,0.646,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019


In [157]:
#a cópia permanece a mesma
df_produto

,PRODUTO
0,ETANOL HIDRATADO
1,ETANOL HIDRATADO
2,ETANOL HIDRATADO
3,ETANOL HIDRATADO
4,ETANOL HIDRATADO
...,...
106818,GNV
106819,GNV
106820,GNV
106821,GNV


In [158]:
#retornando para valores originais com a Series na qual o copy() foi utilizado

df["PRODUTO"] = df_produto
df

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,0.065,...,3.690,0.818,0.018,2.756,0,2.756,2.756,0,6,2019
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,0.129,...,3.789,0.95,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,0.190,...,3.499,0.914,0.065,1.998,0,1.9981,1.9981,0,6,2019
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,0.229,...,3.490,0.646,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019


#### 1.3.4.3 Criando novas colunas
Para **criar uma nova coluna** em um data frame, basta atribuirmos uma _lista/Series de valores_ a uma **nova 'chave'** do data frame. <br/>

**PS:** A _quantidade de valores_ da lista precisa ser **igual** ao _número de linhas/registros_ do data frame.

In [159]:
#criando coluna com valor constante, aparece no último lugar
df['COL TESTE'] = 'default'
df

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,COL TESTE
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004,default
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004,default
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004,default
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004,default
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004,default
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,0.065,...,0.818,0.018,2.756,0,2.756,2.756,0,6,2019,default
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,0.129,...,0.95,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019,default
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,0.190,...,0.914,0.065,1.998,0,1.9981,1.9981,0,6,2019,default
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,0.229,...,0.646,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019,default


In [160]:
#criando coluna a partir de lista
#len(coluna) rpecisa bater com numero de linhas da tabela

df['Coluna a partir de lista'] = range(df.shape[0])
df

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,COL TESTE,Coluna a partir de lista
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004,default,0
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004,default,1
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004,default,2
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004,default,3
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004,default,4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,0.065,...,0.018,2.756,0,2.756,2.756,0,6,2019,default,106818
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,0.129,...,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019,default,106819
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,0.190,...,0.065,1.998,0,1.9981,1.9981,0,6,2019,default,106820
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,0.229,...,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019,default,106821


In [161]:
#coluna fazendo referencia a outra coluna

df['PRECO EM DOLARES'] = df['PREÇO MÉDIO REVENDA']*6
df

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,COL TESTE,Coluna a partir de lista,PRECO EM DOLARES
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,0.825,0.11,0.4201,0.9666,0.133,5,2004,default,0,7.728
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,0.763,0.088,0.5013,1.05,0.115,5,2004,default,1,6.972
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,0.97,0.095,0.5614,1.161,0.098,5,2004,default,2,8.334
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,0.83,0.119,0.5991,1.22242,0.143,5,2004,default,3,7.572
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,0.941,0.077,0.7441,1.0317,0.082,5,2004,default,4,7.086
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,0.065,...,2.756,0,2.756,2.756,0,6,2019,default,106818,21.444
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,0.129,...,2.451,0.402,1.9842,2.8661,0.164,6,2019,default,106819,20.406
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,0.190,...,1.998,0,1.9981,1.9981,0,6,2019,default,106820,17.472
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,0.229,...,2.374,0.165,2.0179,2.5093,0.07,6,2019,default,106821,18.120


**PS:** Obviamente, a lógica correta em converter o preço dos combustíveis em reais para dólares não é considerar uma taxa de câmbio fixa, uma vez que cada preço foi aferido em um momento diferente.

### 1.3.4 Índices

Todo Data Frame possui **índices**, que não são considerado colunas da tabela. Tais índices são comumente **númericos**, de 0 a num_linhas-1, mas também podem ser **textuais (rótulos/labels)**.

In [162]:
df.index

RangeIndex(start=0, stop=106823, step=1)

Use `list(data.index)` ou `data.index.to_list()` para converter um RangeIndex para uma python list.

#### **Exemplo de Data Frame com Índices Textuais (labels)**

In [163]:
pesquisa_de_satisfacao = pd.DataFrame({
    'bom': [50, 21, 100],
    'ruim': [131, 2, 30],
    'pessimo': [30, 20, 1]
}, index=['XboxOne', 'Playstation4', 'Switch'])

In [164]:
pesquisa_de_satisfacao

,bom,ruim,pessimo
XboxOne,50,131,30
Playstation4,21,2,20
Switch,100,30,1


In [165]:
pesquisa_de_satisfacao.index

Index(['XboxOne', 'Playstation4', 'Switch'], dtype='object')

### 1.4 Selecionando uma ou mais observações (Indexação)

#### **==>  Index-based selection (seleção baseada em Índices)**
Mostrando linhas específicas de um DataFrame:

`iloc`: seleciona elementos do Dataframe, baseado em seu **índice (número)** --> row-first, column-second

**Selecionando uma observação/linha:**

In [166]:
#seleção de uma linha
df.iloc[1]

,1
Unnamed: 0,1
DATA INICIAL,2004-05-09
DATA FINAL,2004-05-15
REGIÃO,CENTRO OESTE
ESTADO,GOIAS
PRODUTO,ETANOL HIDRATADO
NÚMERO DE POSTOS PESQUISADOS,387
UNIDADE DE MEDIDA,R$/l
PREÇO MÉDIO REVENDA,1.162
DESVIO PADRÃO REVENDA,0.114


In [167]:
#seleção de várias linha usando slice
#no caso, de 0 a 10
df.iloc[:10]

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,COL TESTE,Coluna a partir de lista,PRECO EM DOLARES
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,0.825,0.11,0.4201,0.9666,0.133,5,2004,default,0,7.728
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,0.763,0.088,0.5013,1.05,0.115,5,2004,default,1,6.972
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,0.97,0.095,0.5614,1.161,0.098,5,2004,default,2,8.334
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,0.83,0.119,0.5991,1.22242,0.143,5,2004,default,3,7.572
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,0.941,0.077,0.7441,1.0317,0.082,5,2004,default,4,7.086
5,5,2004-05-09,2004-05-15,NORDESTE,BAHIA,ETANOL HIDRATADO,408,R$/l,1.383,0.132,...,0.957,0.128,0.5686,1.35,0.134,5,2004,default,5,8.298
6,6,2004-05-09,2004-05-15,NORDESTE,CEARA,ETANOL HIDRATADO,278,R$/l,1.453,0.218,...,1.1,0.068,0.7332,1.294,0.062,5,2004,default,6,8.718
7,7,2004-05-09,2004-05-15,NORDESTE,MARANHAO,ETANOL HIDRATADO,105,R$/l,1.631,0.158,...,1.116,0.091,0.9923,1.36183,0.082,5,2004,default,7,9.786
8,8,2004-05-09,2004-05-15,NORDESTE,PARAIBA,ETANOL HIDRATADO,125,R$/l,1.284,0.130,...,0.931,0.072,0.7577,1.2972,0.077,5,2004,default,8,7.704
9,9,2004-05-09,2004-05-15,NORDESTE,PERNAMBUCO,ETANOL HIDRATADO,423,R$/l,1.224,0.141,...,0.947,0.091,0.5643,1.3541,0.096,5,2004,default,9,7.344


In [168]:
#seleção de várias linhas, mas não sequenciais, usando listas

df.iloc[[1,5,10,15
        ]]

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,COL TESTE,Coluna a partir de lista,PRECO EM DOLARES
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,0.763,0.088,0.5013,1.05,0.115,5,2004,default,1,6.972
5,5,2004-05-09,2004-05-15,NORDESTE,BAHIA,ETANOL HIDRATADO,408,R$/l,1.383,0.132,...,0.957,0.128,0.5686,1.35,0.134,5,2004,default,5,8.298
10,10,2004-05-09,2004-05-15,NORDESTE,PIAUI,ETANOL HIDRATADO,89,R$/l,1.573,0.134,...,1.079,0.078,0.8857,1.2499,0.072,5,2004,default,10,9.438
15,15,2004-05-09,2004-05-15,NORTE,AMAZONAS,ETANOL HIDRATADO,53,R$/l,1.578,0.123,...,1.137,0.068,0.9515,1.2109,0.06,5,2004,default,15,9.468


In [169]:
#pode ser passado fora de ordem

df.iloc[[1,10,15, 5
        ]]

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,COL TESTE,Coluna a partir de lista,PRECO EM DOLARES
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,0.763,0.088,0.5013,1.05,0.115,5,2004,default,1,6.972
10,10,2004-05-09,2004-05-15,NORDESTE,PIAUI,ETANOL HIDRATADO,89,R$/l,1.573,0.134,...,1.079,0.078,0.8857,1.2499,0.072,5,2004,default,10,9.438
15,15,2004-05-09,2004-05-15,NORTE,AMAZONAS,ETANOL HIDRATADO,53,R$/l,1.578,0.123,...,1.137,0.068,0.9515,1.2109,0.06,5,2004,default,15,9.468
5,5,2004-05-09,2004-05-15,NORDESTE,BAHIA,ETANOL HIDRATADO,408,R$/l,1.383,0.132,...,0.957,0.128,0.5686,1.35,0.134,5,2004,default,5,8.298


In [170]:
#localizando célula especifica
df.iloc[1,5]

'ETANOL HIDRATADO'

**Selecionando múltiplias amostras/linhas:**

#### **==>  Label-based selection (seleção baseadas em Rótulos)**

`loc`: seleciona elementos do Dataframe, baseado em seus **rótulos** --> row-first, column-second

In [171]:
pesquisa_de_satisfacao

,bom,ruim,pessimo
XboxOne,50,131,30
Playstation4,21,2,20
Switch,100,30,1


In [172]:
#localizando usando os indices (mesmo com índices implícitos) = iloc
pesquisa_de_satisfacao.iloc[0,1]

np.int64(131)

In [173]:
#localizando usando os rótulos = loc
pesquisa_de_satisfacao.loc['XboxOne','ruim']

np.int64(131)

In [174]:
#filtrando: todas as linhas, mas só com as colunas 'ruim' e 'péssimo'
#intervalos vazios para selecionar tudo (ou 'desde o começo'; 'até o fim')
pesquisa_de_satisfacao.iloc[:, 1:]

,ruim,pessimo
XboxOne,131,30
Playstation4,2,20
Switch,30,1


In [175]:
#se o df possui indíces, é possível capturar as linhas mesmo com loc
df.head()


,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,COL TESTE,Coluna a partir de lista,PRECO EM DOLARES
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,0.825,0.11,0.4201,0.9666,0.133,5,2004,default,0,7.728
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,0.763,0.088,0.5013,1.05,0.115,5,2004,default,1,6.972
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,0.97,0.095,0.5614,1.161,0.098,5,2004,default,2,8.334
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,0.83,0.119,0.5991,1.22242,0.143,5,2004,default,3,7.572
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,0.941,0.077,0.7441,1.0317,0.082,5,2004,default,4,7.086


In [176]:
df.loc[1]

,1
Unnamed: 0,1
DATA INICIAL,2004-05-09
DATA FINAL,2004-05-15
REGIÃO,CENTRO OESTE
ESTADO,GOIAS
PRODUTO,ETANOL HIDRATADO
NÚMERO DE POSTOS PESQUISADOS,387
UNIDADE DE MEDIDA,R$/l
PREÇO MÉDIO REVENDA,1.162
DESVIO PADRÃO REVENDA,0.114


In [177]:
#mas as colunas precisarão ser pelo nome
df.loc[1, 'PRECO EM DOLARES']

np.float64(6.9719999999999995)

### 1.5 Selecionando um ou mais atributos (colunas)

In [178]:
df.loc[:5, 'ESTADO']

,ESTADO
0,DISTRITO FEDERAL
1,GOIAS
2,MATO GROSSO
3,MATO GROSSO DO SUL
4,ALAGOAS
5,BAHIA


In [179]:
#filtrar colunas: usar lista, retorna na ordem colocada
df[['ESTADO', 'DATA INICIAL']]

,ESTADO,DATA INICIAL
0,DISTRITO FEDERAL,2004-05-09
1,GOIAS,2004-05-09
2,MATO GROSSO,2004-05-09
3,MATO GROSSO DO SUL,2004-05-09
4,ALAGOAS,2004-05-09
...,...,...
106818,RIO GRANDE DO NORTE,2019-06-23
106819,RIO GRANDE DO SUL,2019-06-23
106820,SANTA CATARINA,2019-06-23
106821,SAO PAULO,2019-06-23


Como o rótulo da coluna 'DATA INICIAL' **possui espaço**, não é possível acessá-la como método: `data.DATA INICIAL`

### 1.6 Removendo um Atributo (Coluna) do Data Frame

Como vimos anteriormente, o atributo 'Unnamed: 0' parece ser um **ruído** em nosso dataset. Desta maneira, vamos eliminá-lo,

In [180]:
#deleta coluna a partir do nome
del df['Unnamed: 0']

In [181]:
#deletar múltiplas colunas: um del por linha
del df['COL TESTE']
del df['Coluna a partir de lista']

In [182]:
df.head()

,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,...,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,PRECO EM DOLARES
0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,1.19,...,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004,7.728
1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,0.89,...,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004,6.972
2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,1.18,...,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004,8.334
3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,1.09,...,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004,7.572
4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,1.05,...,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004,7.086


### 1.7 Salvando um Data Frame

Para salvarmos um Data Frame para um **arquivo CSV**, basta usarmos o método `.to_csv`. <br/>
Por padrão, esse método **salva os índices da tabela como uma coluna no CSV**.<br/>
Como no geral tais índices são números de 0 a n-1, não há necessidade para isso (veja que removemos anteriormente a coluna 'Unnamed: 0' que foi justamente esse caso).<br/>
Desta forma, utilize o parâmetro: `index=False`.

Por padrão, o método utilizará a ',' como separador das colunas. Caso queira alterá-lo, utilize o parâmetro `sep`.

In [183]:
# usar função df.to_csv()

### 1.8 Seleção Condicional: Filtrando amostras

Durante nossas análise exploratórias, frequentemente filtraremos nossas amostras, a partir de certas **condições**, para fins de análise mais específica. <br> <br>
Existem algumas maneiras de fazermos tal filtragem. Antes disso, vamos carregar nosso dataset pré-processado que salvamos no item anterior.

Regras: Gasolina comum, preço acima de R$2,00; estar nos estados de SP ou RJ

In [184]:
df.head(1)

,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,...,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,PRECO EM DOLARES
0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,1.19,...,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004,7.728


In [185]:
filtro_1 = (df['ESTADO'] == 'SAO PAULO') | (df['ESTADO'] == 'RIO DE JANEIRO')
filtro_1

,ESTADO
0,False
1,False
2,False
3,False
4,False
...,...
106818,False
106819,False
106820,False
106821,True


In [186]:
#encontrar valores de PRODUTO
df['PRODUTO'].unique()

array(['ETANOL HIDRATADO', 'GASOLINA COMUM', 'GLP', 'GNV', 'ÓLEO DIESEL',
       'ÓLEO DIESEL S10'], dtype=object)

In [187]:
filtro_2 = df['PRODUTO'] == 'GASOLINA COMUM'
filtro_2

,PRODUTO
0,False
1,False
2,False
3,False
4,False
...,...
106818,False
106819,False
106820,False
106821,False


In [94]:
filtro_3 = df['PREÇO MÉDIO REVENDA'] > 2
filtro_3

,PREÇO MÉDIO REVENDA
0,False
1,False
2,False
3,False
4,False
...,...
106818,True
106819,True
106820,True
106821,True


In [189]:
filtro_final = filtro_1 & filtro_2 & filtro_3

#mostrar as linhas do df que correspondem às regras aplicadas nos filtros 1 a 3
df_filtrado = df[filtro_final]

In [192]:
#validando resultados == estados são apenas RJ e SP? Produto apenas Gasolina comum?

print(df_filtrado['ESTADO'].unique())
print(df_filtrado['PRODUTO'].unique())

['RIO DE JANEIRO' 'SAO PAULO']
['GASOLINA COMUM']


In [194]:
#validando resultados == valores apenas acima de R$2?
df_filtrado['PREÇO MÉDIO REVENDA'].min()

2.028

Alternativa mais performática: salva df temporarios para que as filtragens ocorram em uma quantidade menor de linhas



In [195]:
filtro_1 = (df['ESTADO'] == 'SAO PAULO') | (df['ESTADO'] == 'RIO DE JANEIRO')
postos_sprj = df[filtro_1]

filtro_2 = df["PRODUTO"] == 'GASOLINA COMUM'
produto_gasolinacomum_sprj = postos_sprj[filtro_2]

filtro_3 = df['PREÇO MÁXIMO REVENDA'] > 2
produto_acima2reais_gasolina_comum_sprj = produto_gasolinacomum_sprj[filtro_3]



/tmp/ipython-input-1783691464.py:5: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  produto_gasolinacomum_sprj = postos_sprj[filtro_2]
/tmp/ipython-input-1783691464.py:8: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  produto_acima2reais_gasolina_comum_sprj = produto_gasolinacomum_sprj[filtro_3]


In [196]:
produto_acima2reais_gasolina_comum_sprj

,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,...,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,PRECO EM DOLARES
12086,2004-05-09,2004-05-15,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,1071,R$/l,2.045,0.109,1.700,...,0.254,0.053,1.791,0.047,1.6363,1.9594,0.026,5,2004,12.270
12087,2004-05-09,2004-05-15,SUDESTE,SAO PAULO,GASOLINA COMUM,4107,R$/l,1.891,0.101,1.570,...,0.235,0.053,1.656,0.064,1.38,1.944,0.039,5,2004,11.346
12113,2004-05-16,2004-05-22,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,1055,R$/l,2.040,0.113,1.700,...,0.252,0.055,1.788,0.049,1.5943,1.99651,0.027,5,2004,12.240
12114,2004-05-16,2004-05-22,SUDESTE,SAO PAULO,GASOLINA COMUM,4133,R$/l,1.888,0.101,1.540,...,0.233,0.053,1.655,0.063,1.38,1.9319,0.038,5,2004,11.328
12140,2004-05-23,2004-05-29,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,1091,R$/l,2.042,0.108,1.700,...,0.252,0.053,1.79,0.046,1.63596,2.2137,0.026,5,2004,12.252
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106479,2019-06-09,2019-06-15,SUDESTE,SAO PAULO,GASOLINA COMUM,1643,R$/l,4.210,0.177,3.759,...,0.425,0.042,3.785,0.096,3.52,4.1007,0.025,6,2019,25.260
106623,2019-06-16,2019-06-22,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,418,R$/l,4.927,0.160,4.599,...,0.544,0.032,4.383,0.1,4.1972,4.6239,0.023,6,2019,29.562
106629,2019-06-16,2019-06-22,SUDESTE,SAO PAULO,GASOLINA COMUM,1635,R$/l,4.168,0.177,3.679,...,0.443,0.042,3.725,0.096,3.502,4.0892,0.026,6,2019,25.008
106772,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,422,R$/l,4.928,0.165,4.599,...,0.538,0.033,4.39,0.097,4.2,4.618,0.022,6,2019,29.568


#### **Selecionando apenas os preços dos Postos de São Paulo**

##### **==> Alternativa 1: Seleção Condicional (Comparações diretas)**

O código abaixo retorna uma ***Series ('array') de booleans***, com o número de linhas (amostras) do Data Frame, que informa os registros de preços dos postos do _estado de São Paulo_ (True).

Para **filtrarmos** os registros de postos do estado de São Paulo:

O resultado é um Data Frame com _apenas_ os registros desejados após a **filtragem**.<br/>
Podemos ainda utilizar o método `loc` para o mesmo fim:

##### **==> Alternativa 2: Utilizando o método `query`**

`query` filtra linhas de um DataFrame baseado em uma **query (pergunta)**.

<br/>

Uma boa prática é **salvar o Data Frame filtrado em uma nova variável**. Isso simplifica a complexidade do código para futuras análise feita para os postos de São Paulo.

#### **Selecionando registros de postos do Rio de Janeiro com Preços acima de 2 reais**

Note que o resultado da seleção continua sendo uma _Series de booleans_ com o _mesmo número de linhas/amostras do DataFrame_, de modo que cada linha possuirá um valor booleano indicando se o posto é do Rio de Janeiro e o preço aferido do combustível é maior que 2 reais (True) ou não.

O símbolo **&** significa **AND** na comparação. Essa nomenclatura do python/pandas é diferente das nomenclaturas tradicionais (&&). <br/>
Similarmente:
- **|** representa o **OR** (não é ||)
- **~** representa o **NOT** (não é !)

Alternativamente, poderíamos usar o método `query` para fazermos tal seleção. Porém, isso não é possível especificamente para esse caso, pois o rótulo da coluna 'PREÇO MÉDIO REVENDA' possui caracteres inválidos para o método (cedilha, acentos)

In [ ]:
# Não funciona
# data.query('ESTADO=="RIO DE JANEIRO" and PREÇO MÉDIO REVENDA > 2')

**Aprofundando mais ainda**

A primeira comparação `(data['ESTADO'] == 'RIO DE JANEIRO')` checa, linha a linha (amostra a amostra) do DataFrame, quais são aquelas cujo o estado é RIO DE JANEIRO. Nenhuma averiguação de preços é feita nesse momento. Como resultado, temos uma Series de booleans que responde **apenas** a essa "pergunta" feita.

A segunda comparação `(data['PREÇO MÉDIO REVENDA'] > 2)` checa, linha a linha (amostra a amostra) do DataFrame, quais são os registro cujo preço do combustível é maior que 2 reais. Note que essa comparação checará os postos de **TODOS os estados**. Como resultado, temos uma Series de booleans que responde **apenas** a essa "pergunta" feita.

Por fim, as duas "perguntas" são unidas pelo AND (&) que retorna a "pergunta completa" que fizemos.

Alguns podem argumentar que tal abordagem é **ineficiente**, uma vez que, para cada condição ("pergunta"), estamos varrendo todas as linhas do DataFrame. <br/>
O Pandas _tenta otimizar_ isso ao máximo por de trás dos panos. Mas, de fato, de tivermos um dataset **muito grande** (centenas de milhares de linhas), tal abordagem se tornará _lenta_.

Assim, poderíamos fazer filtragem com múltiplos condicionais em partes:

#### **Selecionando registros de postos de São Paulo ou do Rio de Janeiro com Gasolina Comum acima de 2 reais**

Podemos fazer a solução do "jeito mais lento", percorrendo o DataFrame inteiro _múltiplas vezes_:

<br/>

Alternativamente:

#### **Selecionando registros dos anos de 2008, 2010 e 2012**

**ALTERNATIVA 1**

In [203]:
selecao_anos = (df['ANO'] == 2008) | (df['ANO'] == 2010) | (df['ANO'] == 2012)
df[selecao_anos]

,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,...,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,PRECO EM DOLARES
5098,2008-01-06,2008-01-12,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,80,R$/l,1.806,0.082,1.709,...,0.29,0.045,1.516,0.061,1.3892,1.6157,0.04,1,2008,10.836
5099,2008-01-06,2008-01-12,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,255,R$/l,1.616,0.100,1.470,...,0.293,0.062,1.323,0.05,1.2242,1.6173,0.038,1,2008,9.696
5100,2008-01-06,2008-01-12,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,142,R$/l,1.581,0.095,1.390,...,0.278,0.060,1.303,0.073,1.14,1.45,0.056,1,2008,9.486
5101,2008-01-06,2008-01-12,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,133,R$/l,1.831,0.046,1.680,...,0.382,0.025,1.449,0.044,1.3247,1.5514,0.03,1,2008,10.986
5102,2008-01-06,2008-01-12,NORDESTE,ALAGOAS,ETANOL HIDRATADO,97,R$/l,1.699,0.098,1.530,...,0.199,0.058,1.5,0.062,1.39,1.7689,0.041,1,2008,10.194
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
55754,2012-12-30,2013-01-05,NORDESTE,RIO GRANDE DO NORTE,GNV,17,R$/m3,1.891,0.015,1.860,...,0.466,0.008,1.425,0.098,1.2657,1.4789,0.069,12,2012,11.346
55755,2012-12-30,2013-01-05,SUL,RIO GRANDE DO SUL,GNV,34,R$/m3,1.980,0.040,1.919,...,0.566,0.020,1.414,0.028,1.3997,1.4608,0.02,12,2012,11.880
55756,2012-12-30,2013-01-05,SUL,SANTA CATARINA,GNV,41,R$/m3,2.003,0.034,1.879,...,-,0.017,-,-,-,-,-,12,2012,12.018
55757,2012-12-30,2013-01-05,SUDESTE,SAO PAULO,GNV,82,R$/m3,1.598,0.199,1.399,...,0.397,0.125,1.201,0.182,0.8354,1.535,0.152,12,2012,9.588


In [204]:
#validando seleção
df[selecao_anos]['ANO'].unique()

array([2008, 2010, 2012])

**ALTERNATIVA 2**

In [205]:
lista_de_anos = [2008,2010,2012]

In [209]:
df['ANO'].isin(lista_de_anos)

,ANO
0,False
1,False
2,False
3,False
4,False
...,...
106818,False
106819,False
106820,False
106821,False


In [211]:
validando_lista = df['ANO'].isin(lista_de_anos)
df[validando_lista]['ANO'].unique()

array([2008, 2010, 2012])

Alternativa 3: usando query


In [212]:
df.query('ANO in @lista_de_anos')

,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,...,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,PRECO EM DOLARES
5098,2008-01-06,2008-01-12,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,80,R$/l,1.806,0.082,1.709,...,0.29,0.045,1.516,0.061,1.3892,1.6157,0.04,1,2008,10.836
5099,2008-01-06,2008-01-12,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,255,R$/l,1.616,0.100,1.470,...,0.293,0.062,1.323,0.05,1.2242,1.6173,0.038,1,2008,9.696
5100,2008-01-06,2008-01-12,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,142,R$/l,1.581,0.095,1.390,...,0.278,0.060,1.303,0.073,1.14,1.45,0.056,1,2008,9.486
5101,2008-01-06,2008-01-12,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,133,R$/l,1.831,0.046,1.680,...,0.382,0.025,1.449,0.044,1.3247,1.5514,0.03,1,2008,10.986
5102,2008-01-06,2008-01-12,NORDESTE,ALAGOAS,ETANOL HIDRATADO,97,R$/l,1.699,0.098,1.530,...,0.199,0.058,1.5,0.062,1.39,1.7689,0.041,1,2008,10.194
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
55754,2012-12-30,2013-01-05,NORDESTE,RIO GRANDE DO NORTE,GNV,17,R$/m3,1.891,0.015,1.860,...,0.466,0.008,1.425,0.098,1.2657,1.4789,0.069,12,2012,11.346
55755,2012-12-30,2013-01-05,SUL,RIO GRANDE DO SUL,GNV,34,R$/m3,1.980,0.040,1.919,...,0.566,0.020,1.414,0.028,1.3997,1.4608,0.02,12,2012,11.880
55756,2012-12-30,2013-01-05,SUL,SANTA CATARINA,GNV,41,R$/m3,2.003,0.034,1.879,...,-,0.017,-,-,-,-,-,12,2012,12.018
55757,2012-12-30,2013-01-05,SUDESTE,SAO PAULO,GNV,82,R$/m3,1.598,0.199,1.399,...,0.397,0.125,1.201,0.182,0.8354,1.535,0.152,12,2012,9.588


### **Iterando com DataFrames**

#### For-each `DataFrame.iterrows()` (LENTO ==> apenas indicado para iterar pequenos conjunto de dados)

In [217]:
#importante! rever isso aqui

for index, row in df.head().iterrows():
  print(f"indice {index} : {row['ESTADO']}")

indice 0 : DISTRITO FEDERAL
indice 1 : GOIAS
indice 2 : MATO GROSSO
indice 3 : MATO GROSSO DO SUL
indice 4 : ALAGOAS


<h2>2. Preparação dos dados</h2>
<hr/>

### 2.1 Removendo amostras com valores vazios (null / nan) no dataset

De um total de 106823 observações, **não há valores null** / nan para nenhum atributo. Mas, veremos que não é bem assim neste caso específico.<br/><br/>

### 2.2 Conversão de tipos de atributos

O pandas automaticamente reconhece os tipos de dados de cada coluna. <br/>
Porém, existem alguns atributos que estão com seus tipos errados: P. ex., "PREÇO MÉDIO DISTRIBUIÇÃO" deveria ser ```float64``` e não ```object```.<br/>
Nestes casos, muito provavelmente algumas amostras têm um string ao invés de um número para tais atributos. <br/>

Os atributos *"DATA INICIAL"* e *"DATA FINAL"* deveriam ser do tipo `datetime`.

Em outros casos, alguns **atributos categóricos** são ```objects```, mas poderiam ter o tipo ```category```, que é um tipo especial do pandas. <br/>
Este tipo é necessário para se utilizar algumas funções específicas do pandas. <br/>
**Não** converteremos para este tipo por ora.

#### **Datas**
Como os atributos de data do datset já estão em um formato de data aceitável (YYYY-MM-DD), não precisamos forçar nenhuma conversão nesse sentido.

#### **Dados Numéricos**

<br/>

Note que temos vários valores ***null*** agora **após a *conversão de tipos***. Vamos checá-los com mais cuidado nos dados originais e preprocessados.

### 2.3 Limpeza de dados

Várias amostras possuem a _string '-'_ em algumas colunas ao invés de um número de fato. Ou seja, não há aferições destes atributos para estas amostras. <br/>

<br/>

Poderíamos **preencher os valores NaN com um valor padrão**. Para isso, basta usar o método `.fillna`.

<br/>

Por mais que a função `fillna` seja interessante e útil em muitos casos, no problema em questão estamos interessados em analisar precisamente, p. ex., o **'PREÇO MÉDIO DISTRIBUIÇÃO'**.<br/>
A fim de não termos valores _sintéticos_ gerados pelo `fillna`, que possam atrapalhar nossa análise, iremos **remover (drop) todas as amostras que possuem qualquer valor NaN** para quaisquer atributos/colunas. <br/>

Para isso, basta utilizarmos o método `dropna`.

Nosso data frame agora, após essa _limpeza_, ficou mais enxuto, contentdo 103392 registros frente aos 106823 registros originais. <br/>

Essas são apenas algumas das possíveis _técnicas de limpeza de dados_. Outras estratégias, p. ex., **confiam na detecção de outliers**, que não veremos neste curso.

#### **Salvando o Dataset Preprocessado**

<h2>3. Estatísticas Descritivas</h2>
<hr/>

O Pandas fornecem algumas funções/métodos ue computam certas estatísticas descritivas.

`describe`: exibe várias **estatísticas descritivas** para os _atributos_ de um dataframe ou para uma _Series_.

<br/>

Como o resultado do `describe` de um _dataframe_ é outro _dataframe_, podemos filtrar apenas algumas colunas.

**Acessando apenas algumas estatísticas**

<br/>

`mean`, `std`, `min`, etc: cada uma das estatísticas do `describe` podem ser computadas individualmente:

#### Qual é o menor preço mínimo de revenda?

#### Qual é a média e desvio padrão dos preços mínimos de revenda?

#### Quais são os estados considerados?

#### Quantos registros (aferições) cada estado possui?

`.value_counts()`:  Conta a frequência dos valores de uma dada variável (de preferência, _categórica_).

<h2>4. Executando funções para cada item de um DataFrame ou Series</h2>
<hr/>

Uma alternativa ao `for-loop` que vimos anteriormente e que é _lento_, é usarmos _funções próprias do pandas_ que **aplicam/mapeiam uma dada função a todos os elementos de um DataFrame ou Series**, retornando novos elementos "transformados".


<img src='https://github.com/julianavtal/pandas-essencial/blob/main/imagens/apply_map_applymap.png?raw=1' width=300/>


Fonte: https://towardsdatascience.com/introduction-to-pandas-apply-applymap-and-map-5d3e044e93ff

In [ ]:
df = pd.DataFrame({ 'A': [1, 2, 3, 4],
                    'B': [10, 20, 30, 40],
                    'C': [100, 200, 300, 400]},
                     index=['Linha 1', 'Linha 2', 'Linha 3', 'Linha 4'])

In [ ]:
df

,A,B,C
Linha 1,1,10,100
Linha 2,2,20,200
Linha 3,3,30,300
Linha 4,4,40,400


`apply()`: usado para aplicar uma função ao longo de um eixo de um DataFrame ou em valores de uma Series.

<img src='https://github.com/julianavtal/pandas-essencial/blob/main/imagens/pandas_axis.jpg?raw=1' width=500/>

Fonte: https://www.allthesnippets.com/browse/pandas/df_axis.html

<img src='https://github.com/julianavtal/pandas-essencial/blob/main/imagens/apply_axis_1.png?raw=1' width=250/>

<img src='https://github.com/julianavtal/pandas-essencial/blob/main/imagens/apply_axis_0.png?raw=1' width=250/>

##### Usando `lambda` functions

<img src='https://github.com/julianavtal/pandas-essencial/blob/main/imagens/apply_axis_1_mean.png?raw=1' width=350/>

<br/>

`applymap()`: usado para aplicar uma função para **cada elemento** (_element-wise_) de um DataFrame.

In [ ]:
df = pd.DataFrame({ 'A': [1, 2, 3, 4],
                    'B': [10, 20, 30, 40],
                    'C': [100, 200, 300, 400]},
                     index=['Linha 1', 'Linha 2', 'Linha 3', 'Linha 4'])
df

,A,B,C
Linha 1,1,10,100
Linha 2,2,20,200
Linha 3,3,30,300
Linha 4,4,40,400


<br/>

`map()`: usado para aplicar uma função para **cada elemento** (_element-wise_) de uma _Series_.

<h2>5. Agrupamento</h2>
<hr/>

`groupby`: Usado para criar **grupo de elementos** (e.x., baseado nos valores de um atributo). <br/>
**Funções** podem então ser aplicadas para os _elementos de cada grupo_, de modo que os **resultados de cada grupo são combinados**.

<br/>

Também podemos ter agrupamentos por mais de um atributo.

<br/>

`.agg`: **agrega (roda)** uma série de funções para os elementos de um dataframe ou de grupos de um dataframe.

In [ ]:
df = pd.DataFrame([[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9],
                   [np.nan, np.nan, np.nan]],
                  columns=['A', 'B', 'C'])
df

NameError: name 'np' is not defined

<h2>6. Ordenação</h2>
<hr/>

In [ ]:
notas = pd.DataFrame({
    'nome': ['João', 'Maria', 'José', 'Alice'],
    'idade': [20, 21, 19, 20],
    'nota_final': [5.0, 10.0, 6.0, 10.0]
})
notas

`.sort_values()`: ordena valores ao longo de um eixo.

Por padrão, o método retorna uma cópia dos dados ordenados em **ordem crescente (ascendente)**. Podemos alterar isso pelo argumento `ascending`.

<br/>

Podemos ordenar a partir de **mais de uma coluna**:

Ordena os registros, primeiramente, pela coluna 'nota_final' em **ordem descrente**. <br/>
Então, reordena os registros _"empatados"_, ou seja, com a **mesma nota final**, em _ordem alfabética_ (ordem crescente).

<br/>

Note que o dataframe original **não foi alterado** após a ordenação.

Para alterá-lo, use o argumento `inplace=True`:

<h2>7. Exercícios</h2>
<hr/>

Vamos aplicar os conceitos que vimos em alguns exercícios. <br/>
Para isso, utilizaremos o dataset de _preços de combustíveis no Brasil_.

Como há apenas medições de janeiro a junho para o ano de 2019, resolvemos **remover os dados** deste ano da análise.

Note que temos um novo dataframe com 99739 linhas, mas com índices fora desse intervalo. <br/>
Acontece que os registros mantiveram seus índices originais após a query. <br/>

Para resetar os índices de _0 a num_linhas-1_, basta usarmos o método `.reset_index()`.

Os índices agora foram **resetados**. Porém, os índices antigos se transformaram em _uma nova coluna_ chamada 'index'. <br/>
Para removê-la durante o _reset_, basta passarmos o argumento `drop=True`.

### 7.1 Qual a proporção de postos pesquisados para cada combustível em cada região

### 7.2 Como os preços da Gasolina Comum em São Paulo variaram em 2018?

#### **Estatísticas Descritivas**

### 7.3 Como os preços da Gasolina Comum e do Etanol em São Paulo variaram em 2018?

<h2>8. Assuntos para continuar os estudos</h2>
<hr/>

- join
- concat
- plot
- data cleaning